x-pack/plugins/security/server/authentication/providers/pki.test.ts

```diff
@@ -0,0 +1,534 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { PeerCertificate, TLSSocket } from 'tls';
+
+jest.mock('net');
+jest.mock('tls');
+
+import { errors } from 'elasticsearch';
+
+import { elasticsearchServiceMock, httpServerMock } from '../../../../../../src/core/server/mocks';
+import { mockAuthenticatedUser } from '../../../common/model/authenticated_user.mock';
+import {
+  MockAuthenticationProviderOptionsWithJest,
+  mockAuthenticationProviderOptionsWithJest,
+} from './base.mock';
+
+import { PKIAuthenticationProvider } from './pki';
+import {
+  ElasticsearchErrorHelpers,
+  ScopedClusterClient,
+} from '../../../../../../src/core/server/elasticsearch';
+import { Socket } from 'net';
+import { getErrorStatusCode } from '../../errors';
+
+interface MockPeerCertificate extends Partial<PeerCertificate> {
+  issuerCertificate: MockPeerCertificate;
+  fingerprint256: string;
+}
+
+function getMockPeerCertificate(chain: string[] | string) {
+  const mockPeerCertificate = {} as MockPeerCertificate;
+
+  (Array.isArray(chain) ? chain : [chain]).reduce(
+    (certificate, fingerprint, index, fingerprintChain) => {
+      certificate.fingerprint256 = fingerprint;
+      certificate.raw = { toString: (enc: string) => `fingerprint:${fingerprint}:${enc}` };
+
+      // Imitate self-signed certificate that is issuer for itself.
+      certificate.issuerCertificate = index === fingerprintChain.length - 1 ? certificate : {};
+
+      return certificate.issuerCertificate;
+    },
+    mockPeerCertificate as Record<string, any>
+  );
+
+  return mockPeerCertificate;
+}
+
+function getMockSocket({
+  authorized = false,
+  peerCertificate = null,
+}: {
+  authorized?: boolean;
+  peerCertificate?: MockPeerCertificate | null;
+} = {}) {
+  const socket = new TLSSocket(new Socket());
+  socket.authorized = authorized;
+  socket.getPeerCertificate = jest.fn().mockReturnValue(peerCertificate);
+  return socket;
+}
+
+describe('PKIAuthenticationProvider', () => {
+  let provider: PKIAuthenticationProvider;
+  let mockOptions: MockAuthenticationProviderOptionsWithJest;
+  beforeEach(() => {
+    mockOptions = mockAuthenticationProviderOptionsWithJest();
+    provider = new PKIAuthenticationProvider(mockOptions);
+  });
+
+  afterEach(() => jest.clearAllMocks());
+
+  describe('`authenticate` method', () => {
```

Comment:
It's implicitly tested here, but do you think we should have a test just for "self signed certs"? We have coverage from some of the tests below, so it might not be necessary.

Reply:
You mean to have exactly the same test as `gets an access token in exchange to peer certificate chain and stores it in the state.`, but with just self-signed cert in the chain?

Reply:
Yeah... it's a flimsy recommendation that I'm not absolutely in-love with...

Reply:
Got it, sure, will add one - it won't hurt :)